File: clib/examples/13_clip_rect.cpp
// 13_clip_rect.cpp - Clip Rectangle Demo
//
// Demonstrates SetClip / ClearClip / GetClip to restrict drawing to sub-regions.
// Three clip windows on screen:
//   - Shape window: lines, circles, triangles clipped at the boundary
//   - Text window: scrolling text clipped by SetClip
//   - Sprite window: code-generated sprite + scaled copies clipped
//
// Learn: SetClip, ClearClip, GetClipX/Y/W/H, how all draw calls respect clip
//
// Compile (Win32): g++ -o 13_clip_rect.exe 13_clip_rect.cpp -mwindows
// Compile (SDL):   g++ -std=c++11 -O2 -o 13_clip_rect 13_clip_rect.cpp -lSDL2

#if defined(_WIN32) && !defined(USE_SDL)
#include "../GameLib.h"
#else
#include "../GameLib.SDL.h"
#endif

#include <math.h>

static int CreateStarSprite(GameLib &game)
{
    int id = game.CreateSprite(16, 16);
    if (id < 0) return -1;
    for (int y = 0; y < 16; y++)
        for (int x = 0; x < 16; x++)
            game.SetSpritePixel(id, x, y, 0x00000000);
    // Simple star shape
    for (int i = 0; i < 16; i++) {
        game.SetSpritePixel(id, 8, i, COLOR_YELLOW);   // vertical bar
        game.SetSpritePixel(id, i, 8, COLOR_YELLOW);   // horizontal bar
        game.SetSpritePixel(id, i, i, COLOR_GOLD);     // diagonal
        game.SetSpritePixel(id, 15 - i, i, COLOR_GOLD); // other diagonal
    }
    // Bright center
    for (int dy = 6; dy <= 10; dy++)
        for (int dx = 6; dx <= 10; dx++)
            game.SetSpritePixel(id, dx, dy, COLOR_WHITE);
    return id;
}

static void DrawWindowBorder(GameLib &game, int x, int y, int w, int h,
                             const char *title, uint32_t borderColor)
{
    game.DrawRect(x - 1, y - 1, w + 2, h + 2, borderColor);
    game.DrawText(x + 6, y + 4, title, COLOR_WHITE);
}

int main()
{
    GameLib game;
    game.Open(640, 480, "13 - Clip Rectangle Demo", true);

    int star = CreateStarSprite(game);

    // Window regions
    const int W1X = 20,  W1Y = 30,  W1W = 280, W1H = 200; // Shapes
    const int W2X = 330, W2Y = 30,  W2W = 290, W2H = 200; // Text
    const int W3X = 20,  W3Y = 260, W3W = 600, W3H = 190; // Sprites

    while (!game.IsClosed()) {
        if (game.IsKeyPressed(KEY_ESCAPE)) break;
        double t = game.GetTime();

        game.Clear(COLOR_RGB(18, 20, 28));
        game.DrawText(20, 8, "13_clip_rect: SetClip restricts all drawing to sub-regions", COLOR_LIGHT_GRAY);

        // === Window 1: Shapes ===
        game.SetClip(W1X, W1Y, W1W, W1H);
        game.FillRect(W1X, W1Y, W1W, W1H, COLOR_RGB(26, 36, 52));

        // Large triangle that extends beyond clip
        game.FillTriangle(W1X + W1W / 2, W1Y - 30,
                          W1X + W1W + 60, W1Y + W1H + 20,
                          W1X - 60, W1Y + W1H + 20,
                          COLOR_ARGB(100, 255, 180, 40));

        // Sweeping lines
        for (int i = 0; i < 6; i++) {
            int sweep = ((int)(t * 120.0) + i * 50) % (W1W + 140);
            int x0 = W1X - 70 + sweep;
            game.DrawLine(x0, W1Y - 10, x0 + 100, W1Y + W1H + 20,
                          COLOR_ARGB(180, 120 + i * 22, 220 - i * 26, 255));
        }

        // Bouncing circle that goes partially out of bounds
        int cx = W1X + W1W / 2 + (int)(cos(t * 1.7) * (W1W / 2 + 30));
        int cy = W1Y + W1H / 2 + (int)(sin(t * 2.1) * (W1H / 2 - 10));
        game.FillCircle(cx, cy, 28, COLOR_ARGB(200, 255, 90, 120));

        // Ellipse crossing boundary
        game.DrawEllipse(W1X + W1W / 2, W1Y + W1H / 2,
                         W1W / 2 + 40, W1H / 3, COLOR_CYAN);

        game.ClearClip();
        DrawWindowBorder(game, W1X, W1Y, W1W, W1H, "Shape Clip", COLOR_WHITE);

        // === Window 2: Text ===
        game.SetClip(W2X, W2Y, W2W, W2H);
        game.FillRect(W2X, W2Y, W2W, W2H, COLOR_RGB(44, 30, 22));

        // Scrolling marquee
        int marqueeX = W2X + W2W - ((int)(t * 100.0) % (W2W + 400));
        game.DrawText(marqueeX, W2Y + 30, "This text scrolls and is clipped at the window boundary...", COLOR_GOLD);

        // Oscillating text
        int oscX = W2X - 80 + ((int)(t * 60.0) % (W2W + 160));
        game.DrawText(oscX, W2Y + 60, "DrawText is also clipped!", COLOR_WHITE);

        // Clip info
        game.DrawPrintf(W2X + 8, W2Y + 100, COLOR_LIGHT_GRAY,
                        "GetClip: %d,%d %dx%d",
                        game.GetClipX(), game.GetClipY(),
                        game.GetClipW(), game.GetClipH());
        game.DrawPrintf(W2X + 8, W2Y + 120, COLOR_LIGHT_GRAY,
                        "Time: %.1f s", t);

        // Tall text block that overflows bottom
        for (int i = 0; i < 8; i++) {
            game.DrawPrintf(W2X + 8, W2Y + 146 + i * 14, COLOR_GRAY,
                            "Line %d - clipped at bottom", i);
        }

        game.ClearClip();
        DrawWindowBorder(game, W2X, W2Y, W2W, W2H, "Text Clip", COLOR_WHITE);

        // === Window 3: Sprites ===
        game.SetClip(W3X, W3Y, W3W, W3H);
        game.FillRect(W3X, W3Y, W3W, W3H, COLOR_RGB(20, 30, 20));

        // Draw a grid of stars at various scales, some crossing boundaries
        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 10; col++) {
                int sx = W3X - 20 + col * 68 + (int)(sin(t + col * 0.7) * 16);
                int sy = W3Y + 10 + row * 64 + (int)(cos(t + row * 1.1) * 10);
                int sz = 24 + row * 16;
                game.DrawSpriteScaled(star, sx, sy, sz, sz);
            }
        }

        // Large sprite crossing right edge
        int bigX = W3X + W3W - 40 + (int)(sin(t * 0.8) * 60);
        game.DrawSpriteScaled(star, bigX, W3Y + W3H / 2 - 40, 80, 80);

        game.ClearClip();
        DrawWindowBorder(game, W3X, W3Y, W3W, W3H, "Sprite Clip", COLOR_WHITE);

        game.DrawText(20, 460, "ESC to quit | All draws are clipped to their window region", COLOR_DARK_GRAY);

        game.Update();
        game.WaitFrame(60);
    }

    game.FreeSprite(star);
    return 0;
}

File: examples/13_clip_rect.py
"""13_clip_rect.py - Clip Rectangle Demo

Demonstrates set_clip / clear_clip / get_clip to restrict drawing to sub-regions.
Three clip windows on screen:
  - Shape window: lines, circles, triangles clipped at the boundary
  - Text window: scrolling text clipped by set_clip
  - Sprite window: code-generated sprite + scaled copies clipped

Learn: set_clip, clear_clip, get_clip, how all draw calls respect clip
"""

import math

import pyezgame as g


def create_star_sprite(game):
    sid = game.create_sprite(16, 16)
    if sid < 0:
        return -1
    for y in range(16):
        for x in range(16):
            game.set_sprite_pixel(sid, x, y, 0x00000000)
    for i in range(16):
        game.set_sprite_pixel(sid, 8, i, g.COLOR_YELLOW)
        game.set_sprite_pixel(sid, i, 8, g.COLOR_YELLOW)
        game.set_sprite_pixel(sid, i, i, g.COLOR_GOLD)
        game.set_sprite_pixel(sid, 15 - i, i, g.COLOR_GOLD)
    for dy in range(6, 11):
        for dx in range(6, 11):
            game.set_sprite_pixel(sid, dx, dy, g.COLOR_WHITE)
    return sid


def draw_window_border(game, x, y, w, h, title, border_color) -> None:
    game.draw_rect(x - 1, y - 1, w + 2, h + 2, border_color)
    game.draw_text(x + 6, y + 4, title, g.COLOR_WHITE)


def main() -> None:
    game = g.GameLib()
    game.open(640, 480, "13 - Clip Rectangle Demo", True)

    star = create_star_sprite(game)

    # Window regions
    w1x, w1y, w1w, w1h = 20, 30, 280, 200  # Shapes
    w2x, w2y, w2w, w2h = 330, 30, 290, 200  # Text
    w3x, w3y, w3w, w3h = 20, 260, 600, 190  # Sprites

    while not game.is_closed():
        if game.is_key_pressed(g.KEY_ESCAPE):
            break
        t = game.get_time()

        game.clear(g.COLOR_RGB(18, 20, 28))
        game.draw_text(20, 8, "13_clip_rect: SetClip restricts all drawing to sub-regions", g.COLOR_LIGHT_GRAY)

        # === Window 1: Shapes ===
        game.set_clip(w1x, w1y, w1w, w1h)
        game.fill_rect(w1x, w1y, w1w, w1h, g.COLOR_RGB(26, 36, 52))

        game.fill_triangle(
            w1x + w1w // 2,
            w1y - 30,
            w1x + w1w + 60,
            w1y + w1h + 20,
            w1x - 60,
            w1y + w1h + 20,
            g.COLOR_ARGB(100, 255, 180, 40),
        )

        for i in range(6):
            sweep = (int(t * 120.0) + i * 50) % (w1w + 140)
            x0 = w1x - 70 + sweep
            game.draw_line(x0, w1y - 10, x0 + 100, w1y + w1h + 20, g.COLOR_ARGB(180, 120 + i * 22, 220 - i * 26, 255))

        cx = w1x + w1w // 2 + int(math.cos(t * 1.7) * (w1w // 2 + 30))
        cy = w1y + w1h // 2 + int(math.sin(t * 2.1) * (w1h // 2 - 10))
        game.fill_circle(cx, cy, 28, g.COLOR_ARGB(200, 255, 90, 120))

        game.draw_ellipse(w1x + w1w // 2, w1y + w1h // 2, w1w // 2 + 40, w1h // 3, g.COLOR_CYAN)

        game.clear_clip()
        draw_window_border(game, w1x, w1y, w1w, w1h, "Shape Clip", g.COLOR_WHITE)

        # === Window 2: Text ===
        game.set_clip(w2x, w2y, w2w, w2h)
        game.fill_rect(w2x, w2y, w2w, w2h, g.COLOR_RGB(44, 30, 22))

        marquee_x = w2x + w2w - (int(t * 100.0) % (w2w + 400))
        game.draw_text(marquee_x, w2y + 30, "This text scrolls and is clipped at the window boundary...", g.COLOR_GOLD)

        osc_x = w2x - 80 + (int(t * 60.0) % (w2w + 160))
        game.draw_text(osc_x, w2y + 60, "DrawText is also clipped!", g.COLOR_WHITE)

        clip_info = game.get_clip()
        game.draw_printf(
            w2x + 8,
            w2y + 100,
            g.COLOR_LIGHT_GRAY,
            f"GetClip: {clip_info[0]},{clip_info[1]} {clip_info[2]}x{clip_info[3]}",
        )
        game.draw_printf(w2x + 8, w2y + 120, g.COLOR_LIGHT_GRAY, f"Time: {t:.1f} s")

        for i in range(8):
            game.draw_printf(w2x + 8, w2y + 146 + i * 14, g.COLOR_GRAY, f"Line {i} - clipped at bottom")

        game.clear_clip()
        draw_window_border(game, w2x, w2y, w2w, w2h, "Text Clip", g.COLOR_WHITE)

        # === Window 3: Sprites ===
        game.set_clip(w3x, w3y, w3w, w3h)
        game.fill_rect(w3x, w3y, w3w, w3h, g.COLOR_RGB(20, 30, 20))

        for row in range(3):
            for col in range(10):
                sx = w3x - 20 + col * 68 + int(math.sin(t + col * 0.7) * 16)
                sy = w3y + 10 + row * 64 + int(math.cos(t + row * 1.1) * 10)
                sz = 24 + row * 16
                game.draw_sprite_scaled(star, sx, sy, sz, sz)

        big_x = w3x + w3w - 40 + int(math.sin(t * 0.8) * 60)
        game.draw_sprite_scaled(star, big_x, w3y + w3h // 2 - 40, 80, 80)

        game.clear_clip()
        draw_window_border(game, w3x, w3y, w3w, w3h, "Sprite Clip", g.COLOR_WHITE)

        game.draw_text(20, 460, "ESC to quit | All draws are clipped to their window region", g.COLOR_DARK_GRAY)

        game.update()
        game.wait_frame(60)

    game.free_sprite(star)


if __name__ == "__main__":
    main()
